test/python/transpiler/test_remove_final_measurements.py

```diff
@@ -0,0 +1,316 @@
+        dag = RemoveFinalMeasurements().run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+        self.assertEqual(dag, expected_dag())
+
+    def test_multi_reg_shared_bits_removed(self):
+        """All registers sharing removed bits should be removed."""
+
+        def expected_dag():
+            q0 = QuantumRegister(2, "q0")
+            qc = QuantumCircuit(q0)
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(2, "q0")
+        c0 = ClassicalRegister(2, "c0")
+        qc = QuantumCircuit(q0, c0)
+
+        # Create reg with shared bits (same as c0)
+        c1 = ClassicalRegister(name="c1", bits=qc.clbits)
+        qc.add_register(c1)
+
+        # measure into all clbits of c0
+        qc.measure(0, c0[0])
+        qc.measure(1, c0[1])
+
+        dag = circuit_to_dag(qc)
+        dag = RemoveFinalMeasurements().run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+        self.assertEqual(dag, expected_dag())
+
+    def test_final_measures_share_dest(self):
+        """Multiple final measurements use the same clbit."""
+
+        def expected_dag():
+            qc = QuantumCircuit(QuantumRegister(2, "q0"))
+            return circuit_to_dag(qc)
+
+        rq = QuantumRegister(2, "q0")
+        rc = ClassicalRegister(1, "c0")
+        qc = QuantumCircuit(rq, rc)
+
+        qc.measure(0, 0)
+        qc.measure(1, 0)
+
+        dag = circuit_to_dag(qc)
+        dag = RemoveFinalMeasurements().run(dag)
+
+        self.assertEqual(dag, expected_dag())
+
+    def test_remove_chained_final_measurements(self):
+        """Remove successive final measurements."""
+
+        def expected_dag():
+            q0 = QuantumRegister(1, "q0")
+            q1 = QuantumRegister(1, "q1")
+            c0 = ClassicalRegister(1, "c0")
+            qc = QuantumCircuit(q0, c0, q1)
+
+            qc.measure(q0, c0)
+            qc.measure(q0, c0)
+            qc.barrier()
+            qc.h(q1)
+
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(1, "q0")
+        q1 = QuantumRegister(1, "q1")
+        c0 = ClassicalRegister(1, "c0")
+        c1 = ClassicalRegister(1, "c1")
+        qc = QuantumCircuit(q0, c0, q1, c1)
+
+        qc.measure(q0, c0)
+        qc.measure(q0, c0)
+        qc.barrier()
+        qc.h(q1)
+        qc.measure(q1, c1)
+        qc.measure(q0, c1)
+
+        dag = circuit_to_dag(qc)
+        dag = RemoveFinalMeasurements().run(dag)
+
+        self.assertEqual(dag, expected_dag())
+
+    def test_remove_clbits_without_register(self):
+        """clbits of final measurements not in a register are removed."""
+
+        def expected_dag():
+            q0 = QuantumRegister(1, "q0")
+            qc = QuantumCircuit(q0)
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(1, "q0")
+        qc = QuantumCircuit(q0)
+
+        # Add clbit without adding register
+        clbit = Clbit(ClassicalRegister(1), 0)
```

Comment:
Is there a reason to specify a register at all here?  `Clbit()` is fine, but I'm not sure if the test needed something else.

Reply:
I saw some comments around the codebase referring to "new-style" bits, which made me worry that such bits might trigger a special code path, and consequently might prevent me from testing the right things. It seems like that's not the case, so I can change this.

Reply:
Fixed in f717f58.